README.md
# barrage_flutter
This is a barrage Engine for flutter.

[中文说明](https://github.com/Dorbmon/barrage_flutter/blob/master/README_CN.md)
## How to use
Use MaterialBarrageWidget to Containe your Widget and the barrage will appear over it.

And you need to figure out a BarrageController for MaterialBarrageWidget where you can add your own barrage.

Here is an example.It builds two barrages that have different speeds.We use vector to figure the speed.
```
import 'package:barrage_flutter/barrage_flutter.dart';
import 'package:flutter/material.dart';
void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final BarrageController controller = BarrageController();
  @override
  void initState() {
    super.initState();
    Future.delayed(Duration.zero,() {
      debugPrint ("add");
      controller.push_barrage (barrage (
        vector2 (0,0),
        vector2 (10,0),
        Text ("测试一下")
      ));
      controller.push_barrage (barrage (
        vector2 (1,20),
        vector2 (15,0),
        Text ("测试2下",style: TextStyle (
          color: Colors.red
        ),)
      ));
    });
  }
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body:Center(child:MaterialBarrageWidget(
            child: Icon (
              Icons.ac_unit,
              size: 200,
            ),
            controller: controller,
            )));
  }
}


```


CHANGELOG.md
## [0.0.1] - TODO: Add release date.
初始项目
## [0.1.0] - 基本能用的版本
具有初始化功能

## [0.1.1] - 修复bug
修复了之前的bug.

## [0.1.2] - 增加删除功能
支持删除弹幕

lib/barrage_flutter.dart
library barrage_flutter;

import 'package:flutter/widgets.dart';

class Vector2<T extends num> {
  T x, y;
  Vector2(this.x, this.y);
  Vector2<T> operator *(T v) {
    Vector2<T> tmp = Vector2(this.x, this.y);
    tmp.x *= v;
    tmp.y *= v;
    return tmp;
  }

  Vector2<T> operator +(Vector2<T> v) {
    Vector2<T> tmp = Vector2(this.x, this.y);
    tmp.x += v.x;
    tmp.y += v.y;
    return tmp;
  }
}

class Barrage extends StatefulWidget {
  //弹幕对象
  final Vector2<double> pos, speed;
  Function remove;
  final Function onDelete;
  Barrage(this.pos, this.speed, this.child,
      {Duration duration,  this.onDelete}) {
    if (duration != null) {
      Future.delayed(duration, () {
        remove();
      });
    }
  }
  Widget child;
  @override
  State<StatefulWidget> createState() {
    return _BarrageState();
  }
}

const int INF = 1000;

class _BarrageState extends State<Barrage> with SingleTickerProviderStateMixin {
  AnimationController controller;
  Vector2<double> npos;
  GlobalKey key = GlobalKey();
  initState() {
    super.initState();
    npos = widget.pos;
    controller = new AnimationController(
        duration: const Duration(seconds: INF), vsync: this);
    controller.addListener(() {
      setState(() {
        npos = widget.speed * (INF * controller.value) + widget.pos;
      });
    });
    controller.forward();
  }

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Positioned(
      child: widget.child,
      top: npos.y,
      left: npos.x,
    );
  }
}

class BarrageController {
  dynamic pushBarrage;
  dynamic removeBarrage;
  dynamic barrages;
}

class MaterialBarrageWidget extends StatefulWidget {
  final BarrageController controller;
  final List<Barrage> barrages = List();
  final Widget child;
  MaterialBarrageWidget({this.controller, this.child}) {
    controller.barrages = barrages;
  }
  @override
  State<StatefulWidget> createState() {
    return _MaterialBarrageWidgetState();
  }
}

typedef bool FindBarrageFunction(Barrage barrage);

class _MaterialBarrageWidgetState extends State<MaterialBarrageWidget> {
  initState() {
    super.initState();
    widget.controller.pushBarrage = (Barrage v) {
      v.remove = () {
        setState(() {
          widget.barrages.remove(v);
        });
      };
      setState(() {
        widget.barrages.add(v);
      });
    };
    widget.controller.removeBarrage = (FindBarrageFunction f) {
      setState(() {
        List<Barrage> need = List();
        for (var i in widget.barrages) {
          if (f(i)) {
            need.add(i);
          }
        }
        for (var i in need) {
          widget.barrages.remove(i);
        }
      });
    };
  }

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: <Widget>[
        widget.child,
      ]..addAll(widget.barrages),
    );
  }
}


example/lib/main.dart
import 'package:barrage_flutter/barrage_flutter.dart';
import 'package:flutter/material.dart';
void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final BarrageController controller = BarrageController();
  @override
  void initState() {
    super.initState();
    Future.delayed(Duration.zero,() {
      debugPrint ("add");
      controller.pushBarrage (Barrage (
        Vector2 (0,0),
        Vector2 (10,0),
        Text ("测试一下")
      ));
      controller.pushBarrage (Barrage (
        Vector2 (1,20),
        Vector2 (15,0),
        Text ("测试2下",style: TextStyle (
          color: Colors.red
        ),)
      ));
    });
  }
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body:Center(child:MaterialBarrageWidget(
            child: Icon (
              Icons.ac_unit,
              size: 200,
            ),
            controller: controller,
            )));
  }
}
